rl.py

```diff
@@ -182,71 +320,133 @@ def sec_to_hours(seconds):
-               print("[INFO]  >>    algo: ", algo)
-               print("[INFO]  >>    seed: ", seed)
-               print("-------------------------------\n")
-
-               r, final_r = test(test_env, model, idx, n_test=5)
-               final_rewards.append(final_r)
-               rewards.append(r)
-               steps.append(timer_limit*20*(idx+1))
-
-
-               with open("./Results_benchmark/" +  name + "/rewards_"+id+".txt", 'w') as fp:
-                   json.dump([rewards, steps], fp)
-               with open("./Results_benchmark/" +  name + "/final_rewards_"+id+".txt", 'w') as fp:
-                   json.dump([final_rewards, steps], fp)
+    if TRAIN == 'new':
+        while idx < n_epochs:
+            try:
+                print("\n-------------------------------")
+                print(">>>    Start training n°", idx+1)
+                print("[INFO]  >>    time: ", sec_to_hours(time.time()-start_time))
+                print("[INFO]  >>    scene: ", env_name)
+                print("[INFO]  >>    algo: ", algo_name)
+                print("[INFO]  >>    seed: ", seed)
+                print("-------------------------------\n")
+
+                model.learn(total_timesteps=total_timesteps, reset_num_timesteps=False, progress_bar=True, tb_log_name=f"{model_name}", log_interval=1)
+                model.save(f"{model_dir}/{total_timesteps*(idx+1)}")
+
+                print("\n-------------------------------")
+                print(">>>    Start test n°", idx+1)
+                print("[INFO]  >>    scene: ", env_name)
+                print("[INFO]  >>    algo: ", algo_name)
+                print("[INFO]  >>    seed: ", seed)
+                print("-------------------------------\n")
+
+                r, final_r = test(test_env, model, idx, n_steps, n_episode=n_test, render=True)
+                final_rewards.append(final_r)
+                rewards.append(r)
+                steps.append(total_timesteps*(idx+1))
+
+                with open(f"{log_dir}/{model_name}_0/rewards.txt", 'w') as fp:
+                    json.dump([rewards, steps], fp)
+                with open(f"{log_dir}/{model_name}_0/final_rewards.txt", 'w') as fp:
+                    json.dump([final_rewards, steps], fp)
+
+                if r >= best:
+                    print(">>>    Save training n°", idx+1)
+                    model.save(f"{model_dir}/best")
+                    model_timestep = 'best'
+                    best = r
+
+                idx+=1
+
+            except:
+               print("[ERROR]  >> The simulation failed. Restart from previous id.")
+               
+            model.save(f"{model_dir}/latest")
 
-               if r >= best:
-                   print(">>>    Save training n°", idx+1)
-                   model.save("./Results_benchmark/" +  name + "/best")
+        print(">>   End.")
+        print("[INFO]  >>    time: ", sec_to_hours(time.time()-start_time))
+    
+    elif TRAIN == 'continue':
```

Comment:
Nice to add the possibility to continue the training. Do you have a specific initialisation of the *idx* and *best* parameters?

Reply:
You can choose the saved time step you want to continue from with the command line argument "-ms", but the default is the latest time step in the case of continuing the training, and the best in the case of testing the model.

I still have a problem reloading the idx from where it previously stopped. I could save it as a parameter in a separate file alongside maybe the model hyperparameters and load it back with the model, but for now it restarts from 0, but in the tensorboard logging it will show the total timesteps.

Reply:
Ok perfect!